docs/arch/002-store-responses-in-query-string.md

```diff
@@ -0,0 +1,61 @@
+
+1. It is difficult to generate URLs with pre-populate responses, for example `/flow-name/y/response1/response2/<pre-populated-response>`. Unless the response is for the first question in a flow, it requires pre-filling responses for all former questions which are unlikely to be known. 
+
+1. Changes to a flow can cause issues with previously cached pages. As URLs are not deterministic for a particular page, changes to a flow can alter which page is displayed. For example, a given flow displays an outcome page for the path `/x/y/z`. A question is removed and the flow now shows an outcome page for the path `/x/y`. The cache would continue to return the old question and users would then be redirected to `/x/y/z`, which is no longer a valid path and could result in an error if it was not also cached.
+
+1. We lose existing responses if the user visits an earlier question. To view a previous or earlier question, responses need to be removed from the URL. This means that it is impossible to change a previous response without re-answering all subsequent questions. This can be a particular burden for users if the flow is long to complete, requires complex input (e.g. dates) or information that isn't memorised.
+
+Additionally, the implementation of storing user responses in a session required the creation of a seperate controller `SessionAnswersController`, separate logic for node resolution and updating the flow. This adds extra burden when maintaining or improving the codebase, as new code may need to accommodate both response storage methods.
+
+## Proposal
+
+Add the ability to store user responses using the URL query string. This will become the default method for storing responses and replace use of the URL path. The ability to store responses in a session cookie will remain and require flow to explicitly enable it within configuration (current behaviour).
+
+To implement this change, we could leverage the existing code (controller, node resolution and flow updates) written for storing responses in a session. This is because the underlying data structure storing the responses is the same i.e. a hash keyed by node name. The only additional logic required is to retrieve and update the parameters in the query string instead from the session.
+
+Example:
+- `flow-name/y` => `flow-name/s/question1`
+- `flow-name/y/response1` => `flow-name/s/question2?question1=response1`
+- `flow-name/y/response1/response2` => `flow-name/s/results?question1=response1&question2=question2`
```

Comment:
Ah yup - this is defo not clear. I'm proposing we use node ID (as defined in the flow definition) in the URL. Here's an example of real Smart Answer which I hope make's it clearer:

- `check-uk-visa/y` => `check-uk-visa/s/nationality`
- `check-uk-visa/y/australia` => `check-uk-visa/s/purpose?nationality=australia`
- `check-uk-visa/y/australia/tourism` => `check-uk-visa/s/result?purpose=tourism&nationality=australia`

I'll update this section to make the intent clearer.

I think using copy such as `question{x}` in URL may make it easier to read and more predictable. However, it doesn't fundamentally solve the problem listed in the previous section, for example `question2` could still potentially return different pages if the flow isn't a linear series of questions.

Reply:
Hopefully clarified [here](https://github.com/alphagov/smart-answers/pull/5212/commits/cc007788200e6084273ef0d767ed40e276dfe5f1)

Reply:
Great thanks. I think the node id is better 👍  

Unlikely edge case alert, are we at any risk of two parameterised node ids conflicting? a scenario that comes to mind is a question suffixed with a question mark and a result that is the same name without the question mark.

Reply:
Yup a possibility, we'd have to make sure id's are unique within a flow. Some thing we need to keep in mind when we migrate existing flows.

Reply:
Cool - I guess we might want to have the flow presenter be aware of paths anyway as they get more use in questions and answers. So it could be that a Flow raises an error if you define a colliding name. I guess without the flow being aware of the paths we'd need a bunch of `.gsub(/_/, "-")` and `parameterize` scattered around the code.

Reply:
For sure, probably something like former would be ideal - raising an error when reading in the initial flow definition.